Reef 1448 Support for upstream heartbeats (job cancellation)

andrey-me opened the pull request:
[REEF-1448] Support for upstream heartbeats
This closes JIRA [REEF-148](https://issues.apache.org/jira/browse/REEF-JIRA_ISSUE_NUMBER)
Adding job cancellation functionality into IMRU driver.
And exposing ICancellationDetector interface in API which enables client to provide custom cancellation detection from upstream.
Implementation: JobLifecycleManager is registered as IjobCancelled event source.
the manager checks IsJobCancelled method of configured CancellationonDetector on each timer event and
if cancellation signal is detected JobLifecycleManager fires JobCancelled event.
IMRU driver is registered as IJobCancelled observable and if event is fired, shuts down all evaluators and moves to Fail state.
Similar to failure handling, when job is cancelled, IMRU driver fires and exception instead of retrying the job. (currently there is no other way to communicate driver status)

Client can provide implementation of IJobCancelledDetector in job definition.
This configuration is optional, if no implementation is configured in the definition, the driver assumes there is no cancellation signal.

JIRA:
  [REEF-1448](https://issues.apache.org/jira/browse/REEF-1448)

markusweimer commented:
@andrey-me Can you please squash all the commits into one against current `master`?

andrey-me commented:
yes, done

tcNickolas commented:
Please make sure your change builds successfully. Currently it fails build in AppVeyor due to some StyleCop violations

andrey-me commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs:
```diff
@@ -0,0 +1,192 @@
+    /// if cancellation signal is detected, the manager creates JobCancelled event and propagates the event to all subscriber.
+    /// the manager is used by IMRU driver to enable job cancellation based on job definition.
+    /// </summary>
+    public class JobLifeCycleManager :
```
yes, if we decide to keep the pulling design. (is this good place discuss this?)
The class essentially pulls cancellation signal from IJobCancellationDetector on timer. 
Main reason I added it into driver project because, (imho) most of the clients will use external dependencies to detect cancellation and will want to implement some kind of pulling anyway, so this class makes implementation easier on the client side. Also this timer-based lifecycle manager could be useful to implement other scenarios in driver. For instance: job suspend/resume, stage timeouts ((hence generic name for the class)). 
if we want reef driver configuration to be purely event based, this implementation can be moved to the client, if we move this class out, we can expose Observable<IJobCanceled> in the job definition and have client do the pulling or listening. 
Thoughts?

andrey-me replied:
yes, if we keep this as pull implementation
I added JobLifecycleManager class which essentially pulls cancellation signal from IJobCancellationDetector on timer.
Main reason I added it into driver project because, (imho) most of the clients will use external dependencies to detect cancellation and will want to implement some kind of pulling anyway, so this class makes implementation easier on the client side. Also this timer-based lifecycle manager could be useful to implement other scenarios in driver. For instance: job suspend/resume, stage timeouts ((hence generic name for the class)).
if we want keep driver configuration to be purely event driven, this implementation can be moved to the client: we can expose Observable<IJobCancelled> in the job definition and have client do the pulling implementation.
Thoughts?

---
In reply to: [84164147](https://github.com/apache/reef/pull/1161#discussion_r84164147) [](ancestors = 84164147)

andrey-me commented on lang/cs/Org.Apache.REEF.Tang/Implementations/Configuration/ConfigurationBuilderImpl.cs:
```diff
@@ -85,6 +85,16 @@ protected ConfigurationBuilderImpl(IConfiguration[] confs) : this(new string[0],
         {
         }
 
+        public IConfigurationBuilder AddOptionalConfiguration(IConfiguration conf)
```
Agree, deserves separate decision.
created JIRA to discuss: [REEF-1646] (https://issues.apache.org/jira/browse/REEF-1646)

andrey-me replied:
removed it 

---
In reply to: [84164583](https://github.com/apache/reef/pull/1161#discussion_r84164583) [](ancestors = 84164583)

andrey-me commented:
Yes, will do....I had couple hiccups with rebasing and the build got away from me.

---
In reply to: [255008627](https://github.com/apache/reef/pull/1161#issuecomment-255008627) [](ancestors = 255008627)

andrey-me commented:
Ø  The presence of the new interface IJobCancelledDetector suggests a polling system to detect cancellation. Shouldn't it be event-driven all the way?
Need to close on this before making other changes.
The driver design is still event driven (IObserver<IJobCancelled>) .
But yes, there is pull implementation as well.
I added JobLifecycleManager class which essentially pulls cancellation signal from IJobCancellationDetector on timer.
Main reason I added it into driver project because, (imho) most of the clients will use external dependencies to detect cancellation and will want to implement some kind of pulling anyway, so this class makes implementation easier on the client side. Also this timer-based lifecycle manager could be useful to implement other scenarios in driver. For instance: job suspend/resume, stage timeouts ((hence generic name for the class)).
if we want keep driver configuration to be purely event driven, this implementation can be moved to the client: we can expose Observable<IJobCancelled> in the job definition and have client do the pulling implementation.
Thoughts?

/Andrey

From: Markus Weimer [mailto:notifications@github.com]
Sent: Wednesday, October 19, 2016 2:22 PM
To: apache/reef <reef@noreply.github.com>
Cc: Andrey Meleshko <andreym@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [apache/reef] Reef 1448 (#1161)


@markusweimer requested changes on this pull request.

I've done a first level pass. A couple of things stood out:

  *   This change adds a few new public classes and interfaces. For each, it would be good to check whether they have to be public or whether we can make them internal instead.
  *   The presence of the new interface IJobCancelledDetector suggests a polling system to detect cancellation. Shouldn't it be event-driven all the way?
  *   There are some changes that seem orthogonal to this JIRA. Maybe move them to another one.

________________________________

In lang/cs/Org.Apache.REEF.Driver/JobCancelled.cs<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Freef%2Fpull%2F1161%23pullrequestreview-4958575&data=02%7C01%7Candreym%40microsoft.com%7C7e9cbcbbefe34c5bb30508d3f86600f7%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636125089396948640&sdata=876Yj1s8Es%2Bbvk76ccu8vNP0%2BPJlCCCwMFXdWibBRO4%3D&reserved=0>:

> +// Unless required by applicable law or agreed to in writing,

+// software distributed under the License is distributed on an

+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY

+// KIND, either express or implied.  See the License for the

+// specific language governing permissions and limitations

+// under the License.

+

+using System;

+

+namespace Org.Apache.REEF.Driver.Bridge.Events

+{

+    /// <summary>

+    /// Job cancelled event.

+    /// Wraps timestamp and cancellation message.

+    /// </summary>

+    public sealed class JobCancelled : IJobCancelled

Does this class need to be public?

________________________________

In lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Freef%2Fpull%2F1161%23pullrequestreview-4958575&data=02%7C01%7Candreym%40microsoft.com%7C7e9cbcbbefe34c5bb30508d3f86600f7%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636125089396948640&sdata=876Yj1s8Es%2Bbvk76ccu8vNP0%2BPJlCCCwMFXdWibBRO4%3D&reserved=0>:

> +using System.Timers;

+using Org.Apache.REEF.Driver;

+using Org.Apache.REEF.Driver.Bridge.Events;

+using Org.Apache.REEF.IMRU.API;

+using Org.Apache.REEF.Tang.Annotations;

+using Org.Apache.REEF.Utilities.Logging;

+

+namespace Org.Apache.REEF.IMRU.OnREEF.Driver

+{

+    /// <summary>

+    /// JobLifecycleManager orchestrates job cancellation flow.

+    /// If job cancellation detector is configured in job definition, the manager starts a timer and periodically checks for cancellation signal.

+    /// if cancellation signal is detected, the manager creates JobCancelled event and propagates the event to all subscriber.

+    /// the manager is used by IMRU driver to enable job cancellation based on job definition.

+    /// </summary>

+    public class JobLifeCycleManager :

Can this be internal and sealed?

________________________________

In lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Freef%2Fpull%2F1161%23pullrequestreview-4958575&data=02%7C01%7Candreym%40microsoft.com%7C7e9cbcbbefe34c5bb30508d3f86600f7%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636125089396948640&sdata=876Yj1s8Es%2Bbvk76ccu8vNP0%2BPJlCCCwMFXdWibBRO4%3D&reserved=0>:

> +    /// </summary>

+    public class JobLifeCycleManager :

+        IDisposable,

+        IJobLifecycleManager,

+        IObserver<IDriverStarted>

+    {

+        private static readonly Logger Logger = Logger.GetLogger(typeof(JobLifeCycleManager));

+

+        private Timer timer;

+        private int timerIntervalSec = 1000;

+        private IJobCancelledDetector CancellationDetector { get; set; }

+        private object lockDispose = new object();

+        private readonly List<IObserver<IJobCancelled>> _observers = new List<IObserver<IJobCancelled>>();

+

+        [Inject]

+        public JobLifeCycleManager(

private

________________________________

In lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Freef%2Fpull%2F1161%23pullrequestreview-4958575&data=02%7C01%7Candreym%40microsoft.com%7C7e9cbcbbefe34c5bb30508d3f86600f7%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636125089396948640&sdata=876Yj1s8Es%2Bbvk76ccu8vNP0%2BPJlCCCwMFXdWibBRO4%3D&reserved=0>:

> +

+            try

+            {

+                isCancelled = CancellationDetector != null && CancellationDetector.IsJobCancelled(out cancellationMessage);

+            }

+            catch (Exception ex)

+            {

+                Logger.Log(Level.Error, "IsCancelled check failed. Exception:{0}", ex);

+                isCancelled = false;

+            }

+

+            return isCancelled;

+        }

+

+        [NamedParameter("Sleep interval for Job lifecycle manager", "lifecyclesleepintervalSec", "15")]

+        public sealed class SleepIntervalParameter : Name<int>

Please put parameters into their own namespace, typically Parameters.

________________________________

In lang/cs/Org.Apache.REEF.Tang/Implementations/Configuration/ConfigurationBuilderImpl.cs<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Freef%2Fpull%2F1161%23pullrequestreview-4958575&data=02%7C01%7Candreym%40microsoft.com%7C7e9cbcbbefe34c5bb30508d3f86600f7%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636125089396948640&sdata=876Yj1s8Es%2Bbvk76ccu8vNP0%2BPJlCCCwMFXdWibBRO4%3D&reserved=0>:

> @@ -85,6 +85,16 @@ protected ConfigurationBuilderImpl(IConfiguration[] confs) : this(new string[0],

         {

         }



+        public IConfigurationBuilder AddOptionalConfiguration(IConfiguration conf)

This change seems orthogonal to the goals of the JIRA, right? Maybe move it to a separate JIRA / commit?

________________________________

In lang/cs/Org.Apache.REEF.Tang/Interface/IConfigurationBuilder.cs<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Freef%2Fpull%2F1161%23pullrequestreview-4958575&data=02%7C01%7Candreym%40microsoft.com%7C7e9cbcbbefe34c5bb30508d3f86600f7%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636125089396948640&sdata=876Yj1s8Es%2Bbvk76ccu8vNP0%2BPJlCCCwMFXdWibBRO4%3D&reserved=0>:

> @@ -22,7 +22,8 @@ namespace Org.Apache.REEF.Tang.Interface

 {

     public interface IConfigurationBuilder

     {

-        void AddConfiguration(IConfiguration c);

+        void AddConfiguration(IConfiguration c);

+        IConfigurationBuilder AddOptionalConfiguration(IConfiguration c);

This change seems orthogonal to the goals of the JIRA, right? Maybe move it to a separate JIRA / commit?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fapache%2Freef%2Fpull%2F1161%23pullrequestreview-4958575&data=02%7C01%7Candreym%40microsoft.com%7C7e9cbcbbefe34c5bb30508d3f86600f7%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636125089396948640&sdata=876Yj1s8Es%2Bbvk76ccu8vNP0%2BPJlCCCwMFXdWibBRO4%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAS46jEolsrT0lc4VdBmUX-4PmwQ47h_kks5q1ooHgaJpZM4KbRuE&data=02%7C01%7Candreym%40microsoft.com%7C7e9cbcbbefe34c5bb30508d3f86600f7%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636125089396948640&sdata=e3lSQn%2BoGXsaRSa9UW%2Bj58qX0isZiS%2BtGpL3WlmvYPQ%3D&reserved=0>.

andrey-me commented:
Addressed most of the comments. 
1 test has failed, but appveyor on my local fork is green: https://ci.appveyor.com/project/andrey-me/reef/build/50-REEF-1448. So looks like flaky test.
For pull/push design question, I left explanation and waiting for feedback.

andrey-me commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs:
```diff
@@ -0,0 +1,192 @@
+        }
+
+        [NamedParameter("Sleep interval for Job lifecycle manager", "lifecyclesleepintervalSec", "15")]
+        public sealed class SleepIntervalParameter : Name<int>
```
done

---
In reply to: [84164424](https://github.com/apache/reef/pull/1161#discussion_r84164424) [](ancestors = 84164424)

jwang98052 commented:
@andrey-me Is it ready for review or you will split it into multiple JIRA?

andrey-me commented:
Yes, ready for review.
please ignore AddOptionalConfiguration code - it is waiting on JIRA 1646 decision.

---
In reply to: [256480648](https://github.com/apache/reef/pull/1161#issuecomment-256480648) [](ancestors = 256480648)

andrey-me commented:
removed optional config code as well, replaced with EmptyConfig initialization instead

---
In reply to: [256482076](https://github.com/apache/reef/pull/1161#issuecomment-256482076) [](ancestors = 256482076,256480648)

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/IMRUJobDefinitionBuilderTests.cs:
```diff
@@ -0,0 +1,81 @@
+        }
+
+        [Fact]
+        public void JobDefinitionBuilder_CancellationConfigIsSetToNull()
```
method name should follow the same style, no _ in the names.

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/IMRUJobDefinitionBuilderTests.cs:
```diff
@@ -0,0 +1,81 @@
+        }
+
+        [Fact]
+        public void JobDefinitionBuilder_SetsJobCancellationConfig()
```
same

andrey-me replied:
fixed

---
In reply to: [85995151](https://github.com/apache/reef/pull/1161#discussion_r85995151) [](ancestors = 85995151)

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/IMRUJobDefinitionBuilderTests.cs:
```diff
@@ -0,0 +1,81 @@
+                .GetField("EmptyConfiguration", BindingFlags.NonPublic | BindingFlags.Static)
+                .GetValue(builder) as IConfiguration;
+
+            Assert.Same(defaultConfig, definition.JobCancelSignalConfiguration);
```
Why no compare with Empty configuration directly?

andrey-me replied:
it's private, not assessable

---
In reply to: [85995264](https://github.com/apache/reef/pull/1161#discussion_r85995264) [](ancestors = 85995264)

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/IMRUJobDefinitionBuilderTests.cs:
```diff
@@ -0,0 +1,81 @@
+
+        private IMRUJobDefinitionBuilder CreateTestBuilder()
+        {
+            var testConfig = TangFactory.GetTang().NewConfigurationBuilder().Build();
```
may be name it as empty config as setting the same config to different configurations looks wired, because Tang doesn't allow duplicated configurations.

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/JobLifecycleManagerTest.cs:
```diff
@@ -0,0 +1,210 @@
+        [Fact]
+        public void JobLifeCyclemanger_SendsJobCancelledEvent()
+        {
+            string expectedMessage = "cancelled";
```
typo canceled

andrey-me commented on lang/cs/Org.Apache.REEF.IMRU.Tests/JobLifecycleManagerTest.cs:
```diff
@@ -0,0 +1,210 @@
+            int signalCheckPeriodSec = 1,
+            int waitForEventPeriodSec = 2)
+        {
+            var manager = Activator.CreateInstance(
```
the constructor is private by prev comments

---
In reply to: [85997021](https://github.com/apache/reef/pull/1161#discussion_r85997021) [](ancestors = 85997021)

jwang98052 replied:
Using injection for JobLifeCycleManager is simpler and also a test to the JobLifeCycleManager injectable constructor.

andrey-me replied:
I don't think injection will work in these tests. I need a preconfigured instance of the detector and as I understand Tang does not allow binding instances of objects.
Also these are unit test, not intended to test injection framework.

jwang98052 replied:
It's fine as it is test. Why not also inject IJobCancelledDetector?

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/JobLifecycleManagerTest.cs:
```diff
@@ -0,0 +1,210 @@
+        /// </summary>
+        private class DetectorStubStatic : IJobCancelledDetector 
+        {
+            private bool IsCancelledResponse { get; set; }
```
may be just use private variable instead of property.

andrey-me replied:
properties in general are better as they allow functionality change in future, so I am using properties everywhere.
is there benefit of not using property?

---
In reply to: [86006571](https://github.com/apache/reef/pull/1161#discussion_r86006571) [](ancestors = 86006571)

jwang98052 replied:
Private properties are less commonly used than private variables, unless you want to wrap some logic inside the get/set methods. 
Property access will be (fractionally) slower as it will call the get/set. 
So if you only have simple get/set, my opinion is private variable is better.

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/JobLifecycleManagerTest.cs:
```diff
@@ -0,0 +1,210 @@
+            private string CancellationMessage { get; set; }
+            private Action ActionOnIsCancelledCall { get; set; }
+
+            public DetectorStubStatic(bool isCancelledResponse, string expectedMessage = null, Action testAction = null)
```
private or internal?

andrey-me replied:
fixed

---
In reply to: [86006967](https://github.com/apache/reef/pull/1161#discussion_r86006967) [](ancestors = 86006967)

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/JobLifecycleManagerTest.cs:
```diff
@@ -0,0 +1,210 @@
+        /// <summary>
+        /// Test helper class to provide predefined cancel signal for testing
+        /// </summary>
+        private class DetectorStubStatic : IJobCancelledDetector 
```
May be give another name like SampleJobCancelledDetector.  What stub and static are for in this class name? Little confusion.

andrey-me replied:
fixed

---
In reply to: [86007396](https://github.com/apache/reef/pull/1161#discussion_r86007396) [](ancestors = 86007396)

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/JobLifecycleManagerTest.cs:
```diff
@@ -0,0 +1,210 @@
+        {
+            public IJobCancelled LastEvent { get; private set; }
+
+            public TestObserver(IObservable<IJobCancelled> eventSource)
```
can be private or internal

andrey-me replied:
fixed

---
In reply to: [86007642](https://github.com/apache/reef/pull/1161#discussion_r86007642) [](ancestors = 86007642)

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/JobLifecycleManagerTest.cs:
```diff
@@ -0,0 +1,210 @@
+    public class JobLifecycleManagerTest
+    {
+        [Fact]
+        public void JobLifeCyclemanger_SendsJobCancelledEvent()
```
Same for method names in this class. No "_"

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs:
```diff
@@ -282,5 +285,10 @@ internal IConfiguration ResultHandlerConfiguration
+        /// <summary>
+        /// Configuration for job cancellation signal implementation
+        /// </summary>
+        internal IConfiguration JobCancelSignalConfiguration { get; private set; }
```
Use _jobCancelSignalConfiguration as an instance variable instead of a property to be consistent with others.

andrey-me replied:
done

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/API/JobCancellationDetectorAlwaysFalse.cs:
```diff
@@ -0,0 +1,38 @@
+    /// <summary>
+    /// Default implementation of IJobCancelledDetector. IsJobCancelled returns false always.
+    /// </summary>
+    internal sealed class JobCancellationDetectorAlwaysFalse : IJobCancelledDetector
```
I  would name it as DefaultJobCancelledDetector

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -161,6 +164,11 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
             _serviceAndContextConfigurationProvider =
                 new ServiceAndContextConfigurationProvider<TMapInput, TMapOutput, TPartitionType>(dataSet, configurationManager);
 
+            if (lifecycleManager != null)
```
As you have default implementations for all related constructor, it should be not null.

jwang98052 replied:
Yes, Tang guarantees as there is a default.

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -447,6 +455,9 @@ public void OnNext(ICompletedTask completedTask)
                         TryRecovery();
                         break;
+                    case SystemState.Fail:
+                        FailAction();
```
Right after we change the system state to Fail, we have called FailAction() in the same lock. So if the system state is already Fail, we don't need to call FailAction() any more.

andrey-me replied:
removed

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -641,6 +653,10 @@ public void OnNext(IFailedTask failedTask)
                             break;
 
+                        case SystemState.Fail:
+                            FailAction();
```
Right after we change the system state to Fail, we have called FailAction() in the same lock. So if the system state is already Fail, we don't need to call FailAction() any more.

andrey-me replied:
removed

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -559,6 +570,7 @@ public void OnNext(IFailedEvaluator failedEvaluator)
                             break;
 
                         case SystemState.Fail:
```
This is not expected case. Because when we set system state to Fail, we throw exception.

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -447,6 +455,9 @@ public void OnNext(ICompletedTask completedTask)
                         _taskManager.RecordCompletedTask(completedTask);
                         TryRecovery();
                         break;
+                    case SystemState.Fail:
```
This is not expected case. Because when we set system state to Fail, we throw exception.

jwang98052 replied:
This block is still not removed.

andrey-me replied:
Are you sure it's unexpected ? as I understand the events are asynchronous. The cancelled event might trigger Fail state, while there are still events from other nodes.

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -641,6 +653,10 @@ public void OnNext(IFailedTask failedTask)
                             TryRecovery();
                             break;
 
+                        case SystemState.Fail:
```
This is not expected case. Because when we set system state to Fail, we throw exception.

andrey-me replied:
removed

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -650,6 +666,13 @@ public void OnNext(IFailedTask failedTask)
 
+        public void OnNext(IJobCancelled value)
+        {
+            _cancelEvent = value;
```
Add lock for the entire method.

andrey-me replied:
done

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobCancelled.cs:
```diff
@@ -0,0 +1,39 @@
+    {
+        internal JobCancelled(DateTime timestamp, string message)
+        {
+            this.Timestamp = timestamp;
```
no this as convention

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs:
```diff
@@ -0,0 +1,186 @@
+    /// <summary>
+    /// JobLifecycleManager orchestrates job cancellation flow.
+    /// If job cancellation detector is configured in job definition, the manager starts a timer and periodically checks for cancellation signal.
+    /// if cancellation signal is detected, the manager creates JobCancelled event and propagates the event to all subscriber.
```
subscribers

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs:
```diff
@@ -0,0 +1,186 @@
+        private static readonly Logger Logger = Logger.GetLogger(typeof(JobLifeCycleManager));
+
+        private Timer timer;
+        private int timerIntervalSec = 1000;
```
const

jwang98052 replied:
1000 will never be used because SleepIntervalParameter has default value set as 15.

andrey-me replied:
how about readonly? modified other variables as well

---
In reply to: [86014995](https://github.com/apache/reef/pull/1161#discussion_r86014995) [](ancestors = 86014995)

andrey-me replied:
yep, fixed

---
In reply to: [86017993](https://github.com/apache/reef/pull/1161#discussion_r86017993) [](ancestors = 86017993)

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs:
```diff
@@ -0,0 +1,186 @@
+        private Timer timer;
+        private int timerIntervalSec = 1000;
+        private IJobCancelledDetector CancellationDetector { get; set; }
+        private object lockDispose = new object();
```
_disposeLock

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs:
```diff
@@ -0,0 +1,186 @@
+            this.timerIntervalSec = sleepIntervalSec;
+        }
+
+        ~JobLifeCycleManager()
```
Why destructor?

andrey-me replied:
to guarantee Dispose() call

---
In reply to: [86015377](https://github.com/apache/reef/pull/1161#discussion_r86015377) [](ancestors = 86015377)

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs:
```diff
@@ -0,0 +1,186 @@
+
+        ~JobLifeCycleManager()
+        {
+            Dispose();
```
The dispose() should be called somewhere, e.g. in IMRUDriver.OnNext(IJobCancelled value).

andrey-me replied:
added generic disposal in the driver

---
In reply to: [86016629](https://github.com/apache/reef/pull/1161#discussion_r86016629) [](ancestors = 86016629)

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs:
```diff
@@ -0,0 +1,186 @@
+
+            public AnonymousDisposable(Action disposeAction)
+            {
+                this.DisposeAction = disposeAction;
```
remove this

andrey-me replied:
fixed

---
In reply to: [86017404](https://github.com/apache/reef/pull/1161#discussion_r86017404) [](ancestors = 86017404)

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/Org.Apache.REEF.IMRU.csproj:
```diff
@@ -57,6 +58,7 @@ under the License.
     <Compile Include="API\IMRUUpdateConfiguration.cs" />
     <Compile Include="API\IUpdateFunction.cs" />
     <Compile Include="API\IPerMapperConfigGenerator.cs" />
+    <Compile Include="API\JobCancellationDetectorAlwaysFalse.cs" />
```
this class should not be under API

andrey-me replied:
hmm, I moved it under driver namespace now, but not sure if it's correct:
 we have circular reference b/w folders now Driver depends on API to use the interface, while API depends on Driver for default implementation ....
it doesn't create problems as all this internal in the same project....but it rubs me in wrong way

---
In reply to: [86018302](https://github.com/apache/reef/pull/1161#discussion_r86018302) [](ancestors = 86018302)

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/Org.Apache.REEF.IMRU.csproj:
```diff
@@ -74,6 +76,9 @@ under the License.
     <Compile Include="OnREEF\Driver\DataLoadingContext.cs" />
     <Compile Include="OnREEF\Driver\EvaluatorManager.cs" />
     <Compile Include="OnREEF\Driver\EvaluatorSpecification.cs" />
+    <Compile Include="OnREEF\Driver\IJobLifeCycleManager.cs" />
```
Do we expect the client to implement IJobLifeCycleManager? If yes, it should be under API folder, otherwise it should be internal.

andrey-me replied:
we have default implementation, but for extensibility I think we should allow client to provide their implementation if needed

---
In reply to: [86018616](https://github.com/apache/reef/pull/1161#discussion_r86018616) [](ancestors = 86018616)

jwang98052 commented on lang/cs/Org.Apache.REEF.Tang/Interface/IConfigurationBuilder.cs:
```diff
@@ -22,7 +22,7 @@ namespace Org.Apache.REEF.Tang.Interface
     public interface IConfigurationBuilder
     {
-        void AddConfiguration(IConfiguration c); 
+        void AddConfiguration(IConfiguration c);
```
revert the change to the entire class.

andrey-me replied:
I thought git would automagically remove it from PR
I'll look into it

---
In reply to: [86018763](https://github.com/apache/reef/pull/1161#discussion_r86018763) [](ancestors = 86018763)

andrey-me replied:
looks like we need to wait for final rebase after post-codereview

jwang98052 commented on lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/ImruJobCancelledTests.cs:
```diff
@@ -0,0 +1,254 @@
+using Org.Apache.REEF.Common.Catalog;
+using Org.Apache.REEF.Driver;
+using Org.Apache.REEF.Driver.Bridge;
+using Org.Apache.REEF.Driver.Bridge.Events;
```
remove unused

andrey-me replied:
fixed

---
In reply to: [86018927](https://github.com/apache/reef/pull/1161#discussion_r86018927) [](ancestors = 86018927)

jwang98052 commented on lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/ImruJobCancelledTests.cs:
```diff
@@ -0,0 +1,254 @@
+
+        /// <summary>
+        /// Simple Type to help with Tang injection when constructing IMRUDriver.
+        /// Cares minimum implmentation to satisfy new driver instance for test scenarios
```
typo implementation

andrey-me replied:
fixed

---
In reply to: [86019354](https://github.com/apache/reef/pull/1161#discussion_r86019354) [](ancestors = 86019354)

jwang98052 commented on lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/ImruJobCancelledTests.cs:
```diff
@@ -0,0 +1,254 @@
+                .GetValue(driver) as SystemStateMachine;
+        }
+
+        private void AssertExceptionThrown<TException>(Action ationWithException,
```
This is a simple way to assert expected exception. 
            Action cancle = () => driver.OnNext(cancelledEvent);
            Assert.Throws<IMRUSystemException>(cancle);

andrey-me replied:
Yes, but it doesn't allow me to verify if exception has expected message (for instance cancellation information)


---
In reply to: [86048625](https://github.com/apache/reef/pull/1161#discussion_r86048625) [](ancestors = 86048625)

andrey-me replied:
in general we may look into improving state communication b/w driver and yarn. just throwing exception is very binary and usually not recommended if the error state can be handled gracefully (exceptions are for unexpected cases).

jwang98052 commented:
I completed a pass.

andrey-me commented on lang/cs/Org.Apache.REEF.IMRU.Tests/JobLifecycleManagerTest.cs:
```diff
@@ -0,0 +1,210 @@
+        /// </summary>
+        private class TestObserver : IObserver<IJobCancelled> 
+        {
+            public IJobCancelled LastEvent { get; private set; }
```
fixed

---
In reply to: [86007613](https://github.com/apache/reef/pull/1161#discussion_r86007613) [](ancestors = 86007613)

andrey-me commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs:
```diff
@@ -0,0 +1,186 @@
+
+        private Timer timer;
+        private int timerIntervalSec = 1000;
+        private IJobCancelledDetector CancellationDetector { get; set; }
```
sure

---
In reply to: [86014685](https://github.com/apache/reef/pull/1161#discussion_r86014685) [](ancestors = 86014685)

andrey-me commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs:
```diff
@@ -0,0 +1,186 @@
+    {
+        private static readonly Logger Logger = Logger.GetLogger(typeof(JobLifeCycleManager));
+
+        private Timer timer;
```
fixed (that's why I like using properties instead :-)

---
In reply to: [86014940](https://github.com/apache/reef/pull/1161#discussion_r86014940) [](ancestors = 86014940)

andrey-me commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs:
```diff
@@ -0,0 +1,186 @@
+            IJobCancelledDetector cancelletionDetector,
+            [Parameter(typeof(SleepIntervalParameter))] int sleepIntervalSec)
+        {
+            this.CancellationDetector = cancelletionDetector;
```
fixed

---
In reply to: [86015172](https://github.com/apache/reef/pull/1161#discussion_r86015172) [](ancestors = 86015172)

andrey-me commented:
addressed all comments now. ready for next review.

---
In reply to: [257735275](https://github.com/apache/reef/pull/1161#issuecomment-257735275) [](ancestors = 257735275)

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -126,6 +126,11 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
         private int _numberOfRetries;
 
+        /// <summary>
+        /// manages lifecycle events for driver, like JobCancelled event.
```
Manages

andrey-me replied:
fixed

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs:
```diff
@@ -0,0 +1,186 @@
+                || _cancellationDetector is JobCancellationDetectorAlwaysFalse
+                || _timerIntervalSec <= 0)
+            {
+                Logger.Log(Level.Info, "Cancellation detector is null or default or timer internval is negative - no need to start Timer for job lifecycle manager. _cancellationDetector: {0}, _timer interval: {1}", _cancellationDetector, _timerIntervalSec);
```
typo internval->interval

andrey-me replied:
fixed

andrey-me replied:
fixed

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs:
```diff
@@ -0,0 +1,186 @@
+        {
+            if (!_observers.Any())
+            {
+                Logger.Log(Level.Warning, "There are no observers for cancellation event: skipping cancellation detection");
```
Level.Info as observer here is not mandatory.

andrey-me replied:
fixed

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -713,10 +743,39 @@ private void DoneAction()
+        /// <summary>
+        /// Dispose resources
+        /// </summary>
+        private void Dispose()
```
To avoid confusing with the Dispose in IDisposable, may be another name?

andrey-me replied:
renamed to DisposeResources()

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -713,10 +743,39 @@ private void DoneAction()
+                    }
+                });
+                _disposableResources.Clear();
+            }
```
Should we also call lifecycleManager.Dispose() here? That would explicitly kill the timer in lifecycleManager before driver shut down.

andrey-me replied:
I don;'t think so, IMRU driver does not create instance of JobLIfecycleManager. it just gets it as dependency.

andrey-me commented on lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/ImruJobCancelledTests.cs:
```diff
@@ -0,0 +1,254 @@
+        {
+            var driver = TangFactory
+                    .GetTang()
+                    .NewInjector(GetDriverConfig<ImruDriverTestType, ImruDriverTestType, ImruDriverTestType, ImruDriverTestType>())
```
this is test code, these types are not used, 
The test is using them to fake driver configuration and run cancellation code path. 

---
In reply to: [86050124](https://github.com/apache/reef/pull/1161#discussion_r86050124) [](ancestors = 86050124)

andrey-me replied:
I am not sure I understand. this is just test code faking driver configuration. these classes are not used. the test just invokes Cancellation code path

---
In reply to: [86050124](https://github.com/apache/reef/pull/1161#discussion_r86050124) [](ancestors = 86050124)

andrey-me commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs:
```diff
@@ -0,0 +1,186 @@
+
+        private void OnTimer(object source, ElapsedEventArgs e)
+        {
+            if (!_observers.Any())
```
fixed

andrey-me commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs:
```diff
@@ -0,0 +1,186 @@
+            }
+            catch (Exception ex)
+            {
+                Logger.Log(Level.Error, "IsCancelled check failed. Exception:{0}", ex);
```
the exception will not do anything as it's on timer thread. 
We could send cancellation signal to the driver and abort the job, but  this might impact lots of users due to a dev error. so I am not sure we should take any drastic steps in this case.

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -161,6 +169,12 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+            if (lifecycleManager != null)
+            {
+                var handle = lifecycleManager.Subscribe(this as IObserver<IJobCancelled>);
+                _disposableResources.Add(handle);
```
Actually does it really matter to remember the handles and call their Dispose() in the end? Because the Dispose() in this case just do unsubscribe. I think the important thing is to call Dispose() of JobLifeCycleManager to make sure Timmer is killed.

andrey-me replied:
changed the logic to stop timer when all observers are gone. For symmetry also ensured that timer starts only when on first subscriber . This removes dependency on DriverStarted event (removed corresponding code)

andrey-me replied:
Not easily, there is some logic in manager which depends on timer lifecycle to be the same as manager lifecycle. We do dispose timer when JobLifecyclemanager is destructed. that should be enough. We can change the design to separate Timer lifecycle from JobManager lifecycle....let me know if feel strongly about it.

jwang98052 replied:
Yes, when the object is out of scope, the destructor will be called.  I just feel it is safer to call the lifecycleManager.Dispose explicitly in IMRUDriver.DisposeResources() as timer is in a separate thread.  As we throw exception in the cancellation case, I am not sure if the system will call all the destructors properly.

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/JobLifecycleManagerTest.cs:
```diff
@@ -0,0 +1,210 @@
+    public class JobLifecycleManagerTest
+    {
+        [Fact]
+        public void JobLifeCyclemangerSendsJobCancelledEvent()
```
document what to test.

andrey-me replied:
done

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/JobLifecycleManagerTest.cs:
```diff
@@ -0,0 +1,210 @@
+        }
+
+        [Fact]
+        public void JobLifeCyclemangerSendsJobCancelledEventToMultiplyObservers()
```
document what to test.

andrey-me replied:
done

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/JobLifecycleManagerTest.cs:
```diff
@@ -0,0 +1,210 @@
+        }
+
+        [Fact]
+        public void JobLifeCyclemangerChecksDetectorPeriodically()
```
document what to test.

andrey-me replied:
done

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/JobLifecycleManagerTest.cs:
```diff
@@ -0,0 +1,210 @@
+        }
+
+        [Fact]
+        public void JobLifeCyclemangerDetectorNullDoesNotSendEvent()
```
document what to test.

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/JobLifecycleManagerTest.cs:
```diff
@@ -0,0 +1,210 @@
+        }
+
+        [Fact]
+        public void JobLifeCyclemangerNoObserversDoesNotCheckForSignal()
```
document what to test.

andrey-me replied:
done

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/API/IMRUJobDefinition.cs:
```diff
@@ -116,6 +118,7 @@ public sealed class IMRUJobDefinition
             _perMapConfigGeneratorConfig = perMapConfigGeneratorConfig;
             _invokeGC = invokeGC;
             _resultHandlerConfiguration = resultHandlerConfiguration;
+            this.JobCancelSignalConfiguration = jobCancelSignalConfiguration;
```
remove this

andrey-me replied:
done

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IJobLifeCycleManager.cs:
```diff
@@ -0,0 +1,32 @@
+using Org.Apache.REEF.Driver;
+using Org.Apache.REEF.Tang.Annotations;
+
+namespace Org.Apache.REEF.IMRU.OnREEF.Driver
```
Should this manager be in API folder? I guess if we want to expose it to the user, it should be in API folder. Otherwise, it should be set as internal.  If we cannot decide it now, change this interface to internal.

andrey-me replied:
not sure why not make it public, but converted to internal. :-)

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs:
```diff
@@ -559,6 +570,7 @@ public void OnNext(IFailedEvaluator failedEvaluator)
                             break;
 
                         case SystemState.Fail:
+                            FailAction();
```
It is still here.

andrey-me replied:
removed

jwang98052 commented:
I completed another pass.

andrey-me commented:
done

---
In reply to: [254892883](https://github.com/apache/reef/pull/1161#issuecomment-254892883) [](ancestors = 254892883)

andrey-me commented on lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/ImruJobCancelledTests.cs:
```diff
@@ -0,0 +1,254 @@
+
+namespace Org.Apache.REEF.Tests.Functional.IMRU
+{
+    public class ImruJobCancelledTests
```
I have another set of functional tests including cancellation, but will move this test and looks into creating simple functional here

andrey-me commented:
addressed all comments. ready for next review cycle.

2 main changes:
- lifecycle manager starts/stops timer based on subscriber count (not on driver started event)
- added fun test based on BroadcastReduce tests.

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/Org.Apache.REEF.IMRU.Tests.csproj:
```diff
@@ -93,6 +94,10 @@ under the License.
       <Project>{79e7f89a-1dfb-45e1-8d43-d71a954aeb98}</Project>
       <Name>Org.Apache.REEF.Utilities</Name>
     </ProjectReference>
+    <ProjectReference Include="..\Org.Apache.REEF.IO\Org.Apache.REEF.IO.csproj">
```
use $(SolutionDir) to replace ..

andrey-me replied:
fixed

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/JobLifeCycleManager.cs:
```diff
@@ -53,34 +52,46 @@ internal sealed class JobLifeCycleManager :
+                return;
+            }
+    
+            Logger.Log(Level.Info, "initializing timer to monitor job status. _timer interval: {0}, cancellation detector: {1}", _timerIntervalSec, _cancellationDetector);
```
What will be logged for _cancellationDetector as an object?

andrey-me replied:
Whatever ToString implementation is. I believe default implementation prints out Type of the object, which gives us information what cancellation detector we are trying to use. I didn't use GetType() as wanted to leave some flexibility to the developer.

jwang98052 commented on lang/cs/Org.Apache.REEF.Tests/Functional/IMRU/IMRUBrodcastReduceTestBase.cs:
```diff
@@ -383,5 +401,29 @@ public void OnNext(IRunningTask value)
+            }
+
+            [NamedParameter("Number of IsCancelled checks before cancelling job", "numberOfChecksBeforeCancelling", "0")]
+            public sealed class NumberOfChecksBeforeCancelling : Name<int>
```
internal?

andrey-me replied:
done

jwang98052 commented on lang/cs/Org.Apache.REEF.Tests/Functional/ReefFunctionalTest.cs:
```diff
@@ -155,7 +155,7 @@ public void Dispose()
         }
 
-        protected void ValidateSuccessForLocalRuntime(int numberOfContextsToClose, int numberOfTasksToFail = 0, int numberOfEvaluatorsToFail = 0, string testFolder = DefaultRuntimeFolder, int retryCount = 60)
+        protected void ValidateSuccessForLocalRuntime(int? numberOfContextsToClose, int numberOfTasksToFail = 0, int numberOfEvaluatorsToFail = 0, string testFolder = DefaultRuntimeFolder, int retryCount = 60, string expectedCancellationMessage = null)
```
This method is mainly for successful cases. If you only want to validate a message, you can just read from log file directly and validate in your test, keeping this method unchanged.

jwang98052 replied:
For example, in your test, 
            string[] lines = ReadLogFile(DriverStdout, "driver", testFolder);
            var expectedCancellationMessageCount = GetMessageCount(lines, expectedCancellationMessage );
            Assert.True(expectedCancellationMessageCount > 1 );

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/JobLifecycleManagerTest.cs:
```diff
@@ -70,6 +73,7 @@ public void JobLifeCyclemangerChecksDetectorPeriodically()
         }
 
         [Fact]
+        [Trait("Description", "Verify that JobLifecycle amanger does not sent any cancellation events if signal is not generated.")]
```
typo amanger

andrey-me replied:
fixed

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/JobLifecycleManagerTest.cs:
```diff
@@ -105,6 +111,35 @@ public void JobLifeCyclemangerNoObserversDoesNotCheckForSignal()
 
+        [Fact]
+        [Trait("Description", "Verify that manager stops checking for cancellation signal after all observers unsubscribed.")]
+        public void JobLifeCyclemangerChecksStopsCancellationChecksWhenAllObserversUnsubscribed()
```
JobLifeCyclemanger -> JobLifeCycleManger...
May be use a shorter name :)

andrey-me replied:
yes, that was confusing name. I rephrased, but kept somewhat long. :-)
I think for test naming is more important even if it's long. (It would be easier to read if we allow underscores in names.)

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/ImruDriverTests.cs:
```diff
@@ -0,0 +1,254 @@
+        /// Simple Type to help with Tang injection when constructing IMRUDriver.
+        /// Cares minimum implementation to satisfy new driver instance for test scenarios
+        /// </summary>
+        public class ImruDriverTestType : IPartitionedInputDataSet, IEvaluatorRequestor
```
May be separate mock impl for each interface would be cleaner.

andrey-me replied:
done

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/ImruDriverTests.cs:
```diff
@@ -0,0 +1,254 @@
+
+namespace Org.Apache.REEF.IMRU.Tests
+{
+    public class ImruDriverTests
```
ImruDriverTestsCancleTests

jwang98052 commented:
I completed this pass.

andrey-me commented on lang/cs/Org.Apache.REEF.IMRU.Tests/ImruDriverTests.cs:
```diff
@@ -0,0 +1,254 @@
+            public IResourceCatalog ResourceCatalog { get; private set; }
+            public void Submit(IEvaluatorRequest request)
+            {
+                // nada
```
fixed better comment :-)

andrey-me commented:
addressed comments. ready for another review cycle

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/ImruDriverTestsCancelTests.cs:
```diff
@@ -43,17 +43,17 @@
 namespace Org.Apache.REEF.IMRU.Tests
 {
-    public class ImruDriverTests
+    public class ImruDriverTestsCancelTests
```
ImruDriverCancelTests

andrey-me replied:
yes, this should be renamed already

jwang98052 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/JobLifecycleManagerTest.cs:
```diff
@@ -113,7 +113,7 @@ public void JobLifeCyclemangerNoObserversDoesNotCheckForSignal()
         [Fact]
         [Trait("Description", "Verify that manager stops checking for cancellation signal after all observers unsubscribed.")]
-        public void JobLifeCyclemangerChecksStopsCancellationChecksWhenAllObserversUnsubscribed()
+        public void JobLifeCyclemangerNoCancellationChecksAfterAllObserversUnsubscribed()
```
JobLifeCycle**M**angerNoCancellationChecksAfterAllObserversUnsubscribed

jwang98052 commented:
I just left a few minor comments. Other than those, everything is good.

andrey-me commented:
made the corrections

jwang98052 commented:
I will test and merge
